challenge_9/rack_request.rb

```diff
@@ -48,25 +52,29 @@
     end
     body << '</ul>'
 
+    # Prepare response
     status = :ok
     headers['Content-Type'] = 'text/html'
-  when ['POST', '/create-post']
-    puts 'Got a new POST request!'
+  elsif request.get? && request.path == '/'
+    body << '<p><strong>Submit a new Blog Post!</p></strong>'
+    body << "<form method='post' enctype='application/x-www-form-urlencoded' action='/create-post'>"
+    body << "<p><label>Blog Title: <input name='title'></label></p>"
+    body << "<p><label>Content: <textarea name='content'></textarea></label></p>"
+    body << '<p><button>Submit post</button></p>'
+    body << '</form>'
 
-    # We no longer need content length, as the StringIO object allows us to read until the EOF without blocking
-    # (Puma handles reading the exact content length from the TCP socket behind the scenes for us).
-    # We know this is a small request, so we can read the entire IO object - however, this could very well be a 
-    # large request (ie. recall file uploads), in which case  we would want to limit the amount of data read at a time
-    # in order to not exhaust our program's memory.
-    # Tom & I explored Puma code here to see how Puma handles reading the body from the TCP Server for large streams of
-    # data (TLDR: Tempfile!)
-    # https://github.com/puma/puma/blob/master/lib/puma/client.rb#L280-L304
-    # https://github.com/puma/puma/blob/7970d14e63836d1c47a086928e533eee766af48d/lib/puma/const.rb#L159-L160
-    message_body = environment['rack.input'].read
+    # Prepare response
+    status = :ok
+    headers['Content-Type'] = 'text/html'
+  elsif request.get?
+    body << "method is #{request.request_method}, path is #{request.path}"
+    status = :ok
+    headers['Content-Type'] = 'text/plain'
+  elsif request.post? && request.path == '/create-post'
+    puts 'Got a new POST request!'
 
-    fields = URI.decode_www_form(message_body)
     post = Blog.new
-    fields.each do |name, value|
+    request.params.each do |name, value|
```

Comment:
This is super neat Tom - thanks for taking the time to share ❤️  
It's pretty wild that Rack is able to parse something like
```
>> parser.parse_nested_query('colours[]=red&colours[]=green&colours[]=blue')
=> {"colours"=>["red", "green", "blue"]}
```
! 

I know we talked previously about the fact that form keys don't have to be unique, so I was wondering how this would get parsed into a hash without the keys clobbering each other. Rack is pretty smart!